File: app/javascript/channels/room_channel.js
// WebSocket処理のクライアントサイドを受け持つファイル

import consumer from "./consumer"

consumer.subscriptions.create("RoomChannel", {
  // 接続時に呼ばれるコールバック
  connected() {
    // Enterを押した時にテキストフィールドの値に対して、speakを実行
    document.
      querySelector('input[data-behavior="room_speaker"]').
      addEventListener('keypress', (event) => {
      if (event.key === 'Enter') {
        this.speak(event.target.value)
        event.target.value = ''
        return event.preventDefault()
      }
    })
  },

  // 接続解除時に呼ばれるコールバック
  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  // サーバーからのデータを受信した時に呼ばれるコールバック
  received(data) {
    // 受け取ったメッセージをHTML上に追加
    const element = document.querySelector('#messages')
    element.insertAdjacentHTML('beforeend', data['message'])
  },

  speak: function (message) {
    // Room channelクラスのspeakメソッドをWebSocket経由で呼び出している
    // サーバーサイドへメッセージを送信
    return this.perform('speak', { message: message });
  }
});

File: app/channels/room_channel.rb
# WebSocket処理のサーバーサイドを受け持つファイル
class RoomChannel < ApplicationCable::Channel
  # 購読後に呼ばれるメソッド
  def subscribed
    # "room_channel"というブロードキャスト用のストリーム名を設定
    # broadcast = 全購読者への送信
    stream_from "room_channel"
  end

  # 購読解除後に呼ばれるメソッド
  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # clientから呼び出された時に呼ばれるメソッド
  def speak(data)
    puts "data"
    puts data
    message = Message.create!(content: data["message"])
    # クライアントからメッセージが届くと、ActionCable.server.broadcastメソッドを使って、ストリーム名"room_channel"で通信しているクライアントすべてに届いたメッセージを送信
    ActionCable.server.broadcast(
      "room_channel", { message: render_message(message) }
    )
  end

  private
  def render_message(message)
    # ApplicationController.renderは、外部のテンプレートのレンダリングを行えます
    ApplicationController.render(
      # view/messages/_message.html.erbをレンダリングしている
      partial: "messages/message",
      locals: { message: message}
    )
  end
end
